### ISSUE TITLE
Emitter AMQP protocol stucked after MQ broker restart

### ISSUE THREAD

**Issue by bkalas** [quarkusio/quarkus#40592]
### Describe the bug

we are using Emitters for sending message to remote MQ broker.
Normally without  any OnOverflow annotation
`  @Inject
    @Channel(RESPONSE_QUEUE)
    Emitter<String> emitter;
    ...
     public void emit(Message<String> toEmit) {
           this.emitter.send(toEmit);
   ...
`
Few times we observed situation that if remote broker was restarted, emiting of messages stopped working.
For few x (~100) messages we didnt get any errror (i guess x= default size of buffer),
then we started to get errror
`
java.lang.IllegalStateException: SRMSG00034: Insufficient downstream requests to emit item
`
This was only solved by quarkus app restart.
Health report for channel was all time showing OK status.

We are also using a lot of consumers  (`@Incoming` ) of amqp messages, this were susccessfully reconnected and continue to work.

Whn i try to use for example
`
@OnOverflow(FAIL, bufferSize = 100)
`
it at least discnnected after error from channel and started reporting KO in health check, but here bufferSize seems to be ignored, so it is not usable.

### Expected behavior

Emitting channel must be resilient against remote broker restart and must continue to emit messages.

### Actual behavior

Sometimes restarts of remote broker was handled ok, but noit always.

### How to Reproduce?

1. must have installed amq broker (in our case Redhat AMQ7.11)
2. quarkus app with Emitter to some queue in remote broker
3. emit messages in some intervlas and restart broker.
4. for me always ~6 restart of broker was enough to simulate this issue, that aftger one of this restarts no new message ws emitted.

### Output of `java -version`

17

### Quarkus version or git rev

2.16.12.Final

**Comment by ozangunalp** [comment]
Looks like an issue with the send-retry mechanism on the AMQP connector. It doesn't reconnect the client and keeps retrying with the previous (unconnected) sender. 

I have a fix in mind but I need to be able to reproduce this scenario (in the test environment) to test it.

**Comment by MikkoKauhanen** [comment]
Hi,

If it helps at all I created quick and dirty test to _try_ to reproduce the issue which you can find from [link to the repo](https://github.com/MikkoKauhanen/amqp-producer). This test uses Quarkus 3.10.1. 

We have same problem in our services that use quarkus version 3.5.3. In our emitters we have `@OnOverflow(OnOverflow.Strategy.UNBOUNDED_BUFFER) ` as a strategy.

Here is also link to my msg about my findings related to this issue: [link to comment](https://github.com/quarkusio/quarkus/issues/40118#issuecomment-2097703998
)

**Comment by ozangunalp** [comment]
@MikkoKauhanen Thanks for this, I'll check it later today or at the beginning of next week.

**Comment by bkalas** [comment]
@ozangunalp  Hi, anything new?

**Comment by MikkoKauhanen** [comment]
Hi, Just wanted to mention that we have now found out two additional issues related to MQ restart. 

In case there are messages created by the emitter to be produced to broker during the MQ disconnect. After the broker is up there are a lot of connections established by the application to broker. And these connections seem to stay alive but I don't know if these are ever used by the application for producing messages anymore.

We noticed this in our dev env where we use micro instance type of Amazon MQ (maximumConnections = 300) where we exceeded the maximumConnections and our services health checks started to fail.

We also saw that connections to broker increased a lot even when there were no messages tried to be produced. This turned to be caused by the `AMQPConnector` Readiness health check that tries to make a connection through `AmqpCreditBasedSender.isConnected()`.

**Comment by cescoffier** [comment]
> We also saw that connections to broker increased a lot even when there were no messages tried to be produced. This turned to be caused by the AMQPConnector Readiness health check that tries to make a connection through AmqpCreditBasedSender.isConnected().

That's expected no? The readiness check verify that the broker is reachable. So, we need to establish connections. Now, it should be only one connection.

**Comment by MikkoKauhanen** [comment]
> > We also saw that connections to broker increased a lot even when there were no messages tried to be produced. This turned to be caused by the AMQPConnector Readiness health check that tries to make a connection through AmqpCreditBasedSender.isConnected().
> 
> That's expected no? The readiness check verify that the broker is reachable. So, we need to establish connections. Now, it should be only one connection.

Yes I can understand that the connection is tried to be established. But it seems to be one new connection for each q/health endpoint call which is done during message broker disconnect/downtime. 

<img width="300" alt="connections" src="https://github.com/quarkusio/quarkus/assets/97296315/535b435c-32d6-41b3-9f6d-11a10cb49105">

**Comment by MikkoKauhanen** [comment]
Hey,

I updated the demo project to have three tests that tries to reproduce the issues that I am aware. 

1. Connection amount is increasing due to the emitted msgs while broker is not available
2. Connection amount is increasing due to health checks done while broker is not available
3. Producer stops producing messages after connection issues to broker.

[Link to repo](https://github.com/MikkoKauhanen/amqp-producer)

**Comment by ozangunalp** [comment]
@bkalas @MikkoKauhanen thanks again for the reproducer. 
I'll spend some more cycles this week to understand this issue.

**Comment by bkalas** [comment]
Hello, is there something new with this issue?

**Comment by webwebow** [comment]
@ozangunalp 
I'd be interested in news about this issue, too

**Comment by nikosri** [comment]
Hey @ozangunalp @cescoffier , any news about this?

**Comment by cescoffier** [comment]
The emitter has a buffer, but when this buffer is full, it drops the messages or throw an exception. You can configure this with the OnOverflow annotation.

It would be great to attach a reproducer, as it seems that it does what's expected.

**Comment by cescoffier** [comment]
Ah, it seems that it does not reconnect / retry. That would be a big or a missing feature. Again, a reproducer would help.

**Comment by Dzela** [comment]
Hello @ozangunalp @cescoffier 
I'm also facing the same issue.
Any news related to this?

**Comment by Vanjas1** [comment]
Hello @ozangunalp and @cescoffier, anything new for this issue?

**Comment by ozangunalp** [comment]
In Reactive Messaging upstream we've included reconnection to send retries with https://github.com/smallrye/smallrye-reactive-messaging/pull/2831. Beware that this can (and will) result in losing the original outbound message order in case of reconnection. 
This is mainly an issue using the Emitter to send events to the broker.

To counter this, we've added the option of `max-inflight-messages` to limit the messages sent to the broker waiting for ack.

**Comment by nikosri** [comment]
Hey @ozangunalp thank you very much! Is there a plan to release this change soon?

**Comment by ozangunalp** [comment]
Actively working on it. I'll make sure to keep here updated.

**Comment by ymajoros** [comment]
We're also affected and have stuck messages in production. Any news?

**Comment by ozangunalp** [comment]
I don't think we can backport it. It included a partial rewrite of the AMQP sender. 
Closing this as the original issue is fixed now.
